src/relay/backend/aot/aot_lower_main.cc

```diff
@@ -0,0 +1,868 @@
+      return_sids_.clear();
+      for (auto sid : sinfo->storage_ids) {
+        return_sids_.push_back(sid);
+      }
+    }
+  }
+
+  /*!
+   * \brief Get the necessary storage for the expression.
+   * \param expr The expression.
+   * \return The corresponding token.
+   */
+  StorageInfo GetStorage(const Expr& expr) {
+    // See through "on_device" calls.
+    Expr true_expr = IgnoreOnDevice(expr);
+    VisitExpr(true_expr);
+    auto it = expr_storage_map_.find(true_expr);
+    ICHECK(it != expr_storage_map_.end()) << "Could not find " << true_expr->GetTypeKey() << " "
+                                          << PrettyPrint(true_expr) << " in storage device map";
+    return it->second;
+  }
+
+  /*!
+   * \brief Create storage for the expression.
+   */
+  void CreateStorage(const ExprNode* op) {
+    Expr expr = GetRef<Expr>(op);
+    return CreateStorage(expr, GetVirtualDevice(expr));
+  }
+
+  /*!
+   * \brief Create storage to hold the result of evaluating \p expr in \p virtual_device.
+   */
+  void CreateStorage(const Expr& expr, const VirtualDevice& virtual_device) {
+    ICHECK(!virtual_device->IsFullyUnconstrained())
+        << "invalid virtual device for expr:" << std::endl
+        << PrettyPrint(expr);
+    std::vector<int64_t> storage_ids;
+    std::vector<VirtualDevice> virtual_devices;
+    std::vector<int64_t> storage_sizes_in_bytes;
+    for (const auto& ttype : FlattenTupleType(expr->checked_type())) {
+      storage_ids.push_back(next_available_sid_++);
+      virtual_devices.push_back(virtual_device);
+      storage_sizes_in_bytes.push_back(GetMemorySizeBytes(ttype->shape, ttype->dtype));
+    }
+    expr_storage_map_[expr] = StorageInfo(std::move(storage_ids), std::move(virtual_devices),
+                                          std::move(storage_sizes_in_bytes));
+  }
+
+  /*! \brief Map between Exprs and StorageInfos */
+  StorageMap expr_storage_map_;
+  /*! \brief The next available storage ID to be used */
+  int next_available_sid_{0};
+  /*! \brief The storage IDs that correspond to return values */
+  std::vector<int> return_sids_;
+};
+
+class AOTMainLowerer : public MixedModeVisitor {
+ public:
+  AOTMainLowerer(tvm::CompilationConfig config, CallType call_type)
+      : config_(config), call_type_(call_type) {}
+
+  IRModule Lower(IRModule mod, String mod_name) {
+    VLOG_CONTEXT << "AOT";
+    IRModule lowered_mod = GetRef<IRModule>(mod.CopyOnWrite());
+
+    auto lowered_main = lowered_mod->Lookup("main");
+    auto lowered_main_func = GetRef<Function>(lowered_main.as<FunctionNode>());
+
+    // Assign StorageInfo to all the Relay exprs
+    ExprAllocator expr_allocator;
+    expr_allocator.Run(lowered_main_func);
+    expr_storage_map_ = expr_allocator.GetStorageMap();
+
+    for (auto input : lowered_main_func->params) {
+      input_vars_.push_back(input);
+      std::string input_name = SanitizeName(input->name_hint());
+      // We dont want the compiler changing input names in the
```

Comment:
Nit: I know it is copy pasted code, but
```suggestion
      // We don't want the compiler changing input names in the
```